src/ddscxx/include/org/eclipse/cyclonedds/core/cdr/basic_cdr_ser.hpp

```diff
@@ -62,27 +61,75 @@ class basic_cdr_stream : public cdr_stream {
  *
  * Aligns the stream to the alignment of type T.
  * Reads the value from the current position of the stream str into
- * toread, will swap bytes if necessary.
+ * toread.
  * Moves the cursor of the stream by the size of T.
  * This function is only enabled for arithmetic types and enums.
  *
  * @param[in, out] str The stream which is read from.
  * @param[out] toread The variable to read into.
  */
 template<typename T, std::enable_if_t<std::is_arithmetic<T>::value && !std::is_enum<T>::value, bool> = true >
-void read(basic_cdr_stream &str, T& toread)
+inline void read(basic_cdr_stream &str, T& toread)
 {
   if (str.abort_status())
     return;
 
   str.align(sizeof(T), false);
 
-  char *cursor = str.get_cursor();
-  assert(cursor);
-  transfer_and_swap(*(reinterpret_cast<const T*>(cursor)), toread, str.swap_endianness());
+  toread = *static_cast<T*>(str.get_cursor());
+
   str.incr_position(sizeof(T));
 }
 
+/**
+ * @brief
+ * Primitive type read function, byteswap version.
+ *
+ * Same as read(), with an additional byteswap at the end.
+ *
+ * @param[in, out] str The stream which is read from.
+ * @param[out] toread The variable to read into.
+ */
+template<typename T, std::enable_if_t<std::is_arithmetic<T>::value && !std::is_enum<T>::value, bool> = true >
+inline void read_swapped(basic_cdr_stream &str, T& toread)
+{
+  read(str, toread);
+
+  byte_swap(toread);
```

Comment:
This is also incorrect. The read checks for `str.abort_status()` and returns, the `byte_swap()` does not.

Reply:
Good point (reviewing clearly catches some things that testing doesn't 😁).

I think it makes sense to distinguish between behaviour with valid inputs and that with invalid inputs: the code that's currently in master doesn't deal gracefully with invalid inputs, and against that background, I find it acceptable to take a PR that dramatically improves the performance for valid inputs without improving the situation with invalid inputs.